source/class/qx/tool/cli/commands/Command.js

```diff
@@ -60,353 +57,129 @@ qx.Class.define("qx.tool.cli.commands.Command", {
-      let qooxdooJsonPath = path.join(process.cwd(), qx.tool.config.Registry.config.fileName);
-      let data = {
-        libraries: [],
-        applications: []
-      };
-      if (await fs.existsAsync(qooxdooJsonPath)) {
-        let qooxdooJson = await qx.tool.utils.Json.loadJsonAsync(qooxdooJsonPath);
-        if (qx.lang.Type.isArray(qooxdooJson.libraries)) {
-          data.libraries = qooxdooJson.libraries;
-        }
-        if (qx.lang.Type.isArray(qooxdooJson.applications)) {
-          data.applications = qooxdooJson.applications;
-        }
-      }
-      if (await fs.existsAsync(path.join(process.cwd(), qx.tool.config.Manifest.config.fileName))) {
-        if (!data.libraries.find(lib => lib.path === ".")) {
-          data.libraries.push({path : "."});
-        }
-      }
-      if (await fs.existsAsync(path.join(process.cwd(), qx.tool.config.Compile.config.fileName))) {
-        if (!data.applications.find(app => app.path === ".")) {
-          data.applications.push({path : "."});
-        }
-      }
-      return data;
+    getArgs() {
+      return this.argv;
     },
 
     /**
-     * Returns the path to the current library. If the current directory contains several libraries,
-     * the first one found is returned.
-     * @throws {Error} Throws an error if no library can be found.
-     * @return {String} A promise that resolves with the absolute path to the library
+     * Check if the current application needs to be migrated
      */
-    getLibraryPath: async function() {
-      let {libraries} = await this.getProjectData();
-      if (libraries instanceof Array && libraries.length) {
-        return path.resolve(process.cwd(), libraries[0].path);
+    async checkMigrations(){
+      let appQxVersion;
+      try {
+        appQxVersion = await this.getAppQxVersion();
+      } catch (e) {
+        // if no application qx verson exists, do nothing
+        return;
       }
-      throw new qx.tool.utils.Utils.UserError("Cannot find library path - are you in the right directory?");
-    },
-
-    /**
-     * Returns the path to the current application, depending on the current
-     * working directory. If a directory contains several applications, the first one found is
-     * returned.
-     * @throws {Error} Throws an error if no application can be found.
-     * @return {Promise<String>} A promise that resolves with the absolute path to the application
-     */
-    getApplicationPath: async function() {
-      let {applications} = await this.getProjectData();
-      if (applications instanceof Array && applications.length) {
-        return path.resolve(process.cwd(), applications[0].path);
+      const semaphore = path.join(process.cwd(), ".qxmigrationcheck");
+      try {
+        await fsp.stat(semaphore);
+        this.debug(`Not checking migration because check is already in progress.`);
+      } catch (e) {
+        // run migration in dry-run mode
+        await fsp.writeFile(semaphore,"");
```

Comment:
We've got a race condition here. Since we're using a semaphore, we're clearly concerned with two processes doing migrations at the same time... but it's possible that the `stat` runs in one of the processes and finds no error, an operating system task switch occurs and `stat` runs in the other process finding no error, and as each process continues, they each enter `writeFile`. In general, it is better not to do a test to see if you can do something and then do it (which causes the race condition), but rather to just "try to do it" and let it fail if it can't.

In this case, that would be done by getting rid of the `stat` and passing options to `writeFile` to tell it to error if the file exists. This sample code shows the general idea: if /tmp/xxx already exists, this will fail.

```
fsp.writeFile("
  /tmp/xxx", 
  "hello world",
  {
    flag : fs.constants.O_WRONLY | fs.constants.O_CREAT | fs.constants.O_EXCL,
    mode: 0o644
  })
 .then(() => console.log("worked"))
 .catch((e) => console.log("failed: " + e));
```

Reply:
You are right in principle, but the use case are only for the case that one `qx` command triggers another one before it has finished, and the chance of a race condition here is almost zero because of the time it takes before a `qx` CLI command actually executes. If this was mission-critical, I'd rewrite it, but it is probably not even real problem - I was already thinking to remove it altogether.